src/common/com/intellij/plugins/haxe/codeInsight/daemon/HaxeProjectSdkSetupValidator.java

```diff
@@ -0,0 +1,169 @@
+    return psiFile != null && psiFile.getLanguage().isKindOf(HaxeLanguage.INSTANCE);
+  }
+
+  @Nullable
+  @Override
+  public String getErrorMessage(@NotNull Project project, @NotNull VirtualFile file) {
+    SdkValidationResult result = validateSdk(project, file);
+    if (result != null) {
+      switch (result) {
+        case MODULE_SDK_NOT_DEFINED:
+          return ProjectBundle.message("module.sdk.not.defined");
+        case PROJECT_SDK_NOT_DEFINED:
+          return ProjectBundle.message("project.sdk.not.defined");
+        case MULTIPLE_ROOTS_FOUND:
+          return HaxeBundle.message("sdk.roots.multiple");
+        case NO_VALID_SDK_ROOTS_FOUND:
+          return HaxeBundle.message("sdk.roots.no.valid.root");
+      }
+    }
+
+    return null;
+  }
+
+  private SdkValidationResult validateSdk(Project project, VirtualFile file) {
+    final Module module = ModuleUtilCore.findModuleForFile(file, project);
+    if (module != null && !module.isDisposed()) {
+      final Sdk sdk = ModuleRootManager.getInstance(module).getSdk();
+      if (sdk == null) {
+        if (ModuleRootManager.getInstance(module).isSdkInherited()) {
+          return PROJECT_SDK_NOT_DEFINED;
+        }
+        else {
+          return MODULE_SDK_NOT_DEFINED;
+        }
+      }
+      else {
+        return validateSdkRoots(sdk);
+      }
+    }
+    return null;
+  }
+
+  private SdkValidationResult validateSdkRoots(Sdk sdk) {
+    List<VirtualFile> roots = getDistinctRoots(sdk);
+
+    if (hasNoValidRoots(roots)) {
+      return NO_VALID_SDK_ROOTS_FOUND;
+    }
+    if (hasMultipleOrEmptyRoots(roots)) {
+      return MULTIPLE_ROOTS_FOUND;
+    }
+
+    return null;
+  }
+
+  private List<VirtualFile> getDistinctRoots(Sdk sdk) {
+    return getDistinctRootsStream(sdk).collect(Collectors.toList());
+  }
+
+  private Stream<VirtualFile> getDistinctRootsStream(Sdk sdk) {
+    return Stream.concat(
+      Arrays.stream(sdk.getRootProvider().getFiles(OrderRootType.CLASSES)),
+      Arrays.stream(sdk.getRootProvider().getFiles(OrderRootType.SOURCES))
+    ).distinct();
+  }
+
+  private boolean hasNoValidRoots(List<VirtualFile> roots) {
+    return roots.stream().noneMatch(root -> root.findChild(STD_TYPES_HX) != null);
+  }
+
+  private boolean hasMultipleOrEmptyRoots(List<VirtualFile> roots) {
+    return roots.size() != 1;
+  }
+
+  @Override
+  public void doFix(@NotNull Project project, @NotNull VirtualFile file) {
+    SdkValidationResult result = validateSdk(project, file);
+    if (result == null) {
+      return;
+    }
+
+    switch (result) {
+      case PROJECT_SDK_NOT_DEFINED:
+      case MODULE_SDK_NOT_DEFINED:
+      case NO_VALID_SDK_ROOTS_FOUND:
+        super.doFix(project, file);
+        break;
+      case MULTIPLE_ROOTS_FOUND:
+        pruneExcessiveRoots(project, file);
+        break;
+    }
+  }
+
+  private void pruneExcessiveRoots(Project project, VirtualFile file) {
```

Comment:
One question about this function: Is it possible that non-Haxe classes (e.g. jar files) are added to classpath entries on the Haxe compiler command line?  If so, we probably still want to leave those in the classpath.

Reply:
Command line is not involved in this process, by this method will be pruned only SDK roots. Source paths from SDK not used explicitly in command line.